File: src-tauri/src/native_video/surface_channel.rs
use std::io;
use std::mem;
use std::os::fd::{FromRawFd, OwnedFd, RawFd};

use serde::{Deserialize, Serialize};

use super::surface_protocol::{SurfaceDescriptor, SurfaceFrameReady, SurfaceFrameRelease};

const MAX_MESSAGE_BYTES: usize = 64 * 1024;
const MAX_MESSAGE_FDS: usize = 5;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SurfaceChannelMessage {
    SurfaceCreated { descriptor: SurfaceDescriptor },
    FrameReady { frame: SurfaceFrameReady },
    FrameRelease { release: SurfaceFrameRelease },
}

#[derive(Debug)]
pub struct ReceivedSurfaceMessage {
    pub message: SurfaceChannelMessage,
    pub fds: Vec<OwnedFd>,
}

pub fn socket_pair() -> io::Result<(OwnedFd, OwnedFd)> {
    let mut sockets = [-1; 2];
    let result = unsafe {
        libc::socketpair(
            libc::AF_UNIX,
            libc::SOCK_SEQPACKET | libc::SOCK_CLOEXEC,
            0,
            sockets.as_mut_ptr(),
        )
    };
    if result != 0 {
        return Err(io::Error::last_os_error());
    }
    Ok(unsafe {
        (
            OwnedFd::from_raw_fd(sockets[0]),
            OwnedFd::from_raw_fd(sockets[1]),
        )
    })
}

pub fn send_message(
    socket: RawFd,
    message: &SurfaceChannelMessage,
    fds: &[RawFd],
) -> io::Result<()> {
    if fds.len() > MAX_MESSAGE_FDS {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "surface message has too many file descriptors",
        ));
    }
    let payload = serde_json::to_vec(message)
        .map_err(|error| io::Error::new(io::ErrorKind::InvalidInput, error))?;
    if payload.len() > MAX_MESSAGE_BYTES {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "surface message exceeds the bounded payload size",
        ));
    }

    let mut iov = libc::iovec {
        iov_base: payload.as_ptr().cast_mut().cast(),
        iov_len: payload.len(),
    };
    let control_bytes = if fds.is_empty() {
        0
    } else {
        unsafe { libc::CMSG_SPACE(mem::size_of_val(fds) as u32) as usize }
    };
    let mut control = vec![0_u8; control_bytes];
    let mut header: libc::msghdr = unsafe { mem::zeroed() };
    header.msg_iov = &mut iov;
    header.msg_iovlen = 1;
    if !control.is_empty() {
        header.msg_control = control.as_mut_ptr().cast();
        header.msg_controllen = control.len();
        unsafe {
            let cmsg = libc::CMSG_FIRSTHDR(&header);
            (*cmsg).cmsg_level = libc::SOL_SOCKET;
            (*cmsg).cmsg_type = libc::SCM_RIGHTS;
            (*cmsg).cmsg_len = libc::CMSG_LEN(mem::size_of_val(fds) as u32) as usize;
            std::ptr::copy_nonoverlapping(
                fds.as_ptr().cast::<u8>(),
                libc::CMSG_DATA(cmsg),
                mem::size_of_val(fds),
            );
        }
    }

    let sent = unsafe { libc::sendmsg(socket, &header, libc::MSG_NOSIGNAL) };
    if sent < 0 {
        return Err(io::Error::last_os_error());
    }
    if sent as usize != payload.len() {
        return Err(io::Error::new(
            io::ErrorKind::WriteZero,
            "surface channel sent a partial packet",
        ));
    }
    Ok(())
}

pub fn receive_message(socket: RawFd) -> io::Result<ReceivedSurfaceMessage> {
    let mut payload = vec![0_u8; MAX_MESSAGE_BYTES];
    let mut iov = libc::iovec {
        iov_base: payload.as_mut_ptr().cast(),
        iov_len: payload.len(),
    };
    let control_bytes =
        unsafe { libc::CMSG_SPACE((MAX_MESSAGE_FDS * mem::size_of::<RawFd>()) as u32) as usize };
    let mut control = vec![0_u8; control_bytes];
    let mut header: libc::msghdr = unsafe { mem::zeroed() };
    header.msg_iov = &mut iov;
    header.msg_iovlen = 1;
    header.msg_control = control.as_mut_ptr().cast();
    header.msg_controllen = control.len();

    let received = unsafe { libc::recvmsg(socket, &mut header, libc::MSG_CMSG_CLOEXEC) };
    if received < 0 {
        return Err(io::Error::last_os_error());
    }
    if received == 0 {
        return Err(io::Error::new(
            io::ErrorKind::UnexpectedEof,
            "surface channel closed",
        ));
    }
    if header.msg_flags & (libc::MSG_TRUNC | libc::MSG_CTRUNC) != 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "surface channel packet or descriptors were truncated",
        ));
    }
    payload.truncate(received as usize);

    let mut raw_fds = Vec::new();
    unsafe {
        let mut cmsg = libc::CMSG_FIRSTHDR(&header);
        while !cmsg.is_null() {
            if (*cmsg).cmsg_level == libc::SOL_SOCKET && (*cmsg).cmsg_type == libc::SCM_RIGHTS {
                let data_bytes = (*cmsg).cmsg_len - libc::CMSG_LEN(0) as usize;
                let count = data_bytes / mem::size_of::<RawFd>();
                let data = libc::CMSG_DATA(cmsg).cast::<RawFd>();
                for index in 0..count {
                    raw_fds.push(*data.add(index));
                }
            }
            cmsg = libc::CMSG_NXTHDR(&header, cmsg);
        }
    }
    if raw_fds.len() > MAX_MESSAGE_FDS {
        for fd in raw_fds {
            unsafe { libc::close(fd) };
        }
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "surface channel received too many descriptors",
        ));
    }

    let message = match serde_json::from_slice(&payload) {
        Ok(message) => message,
        Err(error) => {
            for fd in raw_fds {
                unsafe { libc::close(fd) };
            }
            return Err(io::Error::new(io::ErrorKind::InvalidData, error));
        }
    };
    let fds = raw_fds
        .into_iter()
        .map(|fd| unsafe { OwnedFd::from_raw_fd(fd) })
        .collect();
    Ok(ReceivedSurfaceMessage { message, fds })
}

#[cfg(test)]
mod tests {
    use std::fs::File;
    use std::os::fd::AsRawFd;

    use super::*;
    use crate::native_video::surface_protocol::{SurfaceHandleKind, SurfacePlane};

    #[test]
    fn seqpacket_channel_preserves_one_message_and_ancillary_fd() {
        let (sender, receiver) = socket_pair().unwrap();
        let file = File::open("/dev/null").unwrap();
        let descriptor = SurfaceDescriptor {
            generation: 4,
            buffer_id: 2,
            width: 640,
            height: 360,
            sample_aspect_ratio: 1.0,
            rotation_degrees: 0,
            fourcc: 0x3432_4241,
            modifier: 0,
            handle_kind: SurfaceHandleKind::SharedMemory,
            reusable_dmabuf: false,
            producer_drm_node: None,
            color_space: None,
            color_range: None,
            chroma_location: None,
            planes: vec![SurfacePlane {
                stride: 640 * 4,
                offset: 0,
            }],
            dmabuf: None,
        };
        let message = SurfaceChannelMessage::SurfaceCreated {
            descriptor: descriptor.clone(),
        };
        send_message(sender.as_raw_fd(), &message, &[file.as_raw_fd()]).unwrap();
        let received = receive_message(receiver.as_raw_fd()).unwrap();
        assert_eq!(received.message, message);
        assert_eq!(received.fds.len(), 1);
        assert!(descriptor.validate(received.fds.len()).is_ok());
        assert!(unsafe { libc::fcntl(received.fds[0].as_raw_fd(), libc::F_GETFD) } >= 0);
    }

    #[test]
    fn channel_rejects_more_than_the_bounded_descriptor_count() {
        let (sender, _receiver) = socket_pair().unwrap();
        let file = File::open("/dev/null").unwrap();
        let fds = vec![file.as_raw_fd(); MAX_MESSAGE_FDS + 1];
        let message = SurfaceChannelMessage::FrameRelease {
            release: SurfaceFrameRelease {
                generation: 1,
                buffer_id: 0,
                sequence: 1,
            },
        };
        assert_eq!(
            send_message(sender.as_raw_fd(), &message, &fds)
                .unwrap_err()
                .kind(),
            io::ErrorKind::InvalidInput
        );
    }
}

File: native-video/src/surface_channel.cpp
// SPDX-License-Identifier: MIT
#include "surface_channel.h"

#include <sys/socket.h>
#include <unistd.h>

#include <array>
#include <cerrno>
#include <cstdlib>
#include <cstring>
#include <limits>
#include <stdexcept>
#include <string>
#include <utility>

namespace localbooru::native_video {
namespace {
constexpr std::size_t kMaxMessageBytes = 64U * 1024U;
constexpr std::size_t kMaxMessageFds = 5;

std::runtime_error system_error(const char* action) {
  return std::runtime_error(std::string(action) + ": " + std::strerror(errno));
}

void close_all(std::vector<int>& fds) {
  for (const int fd : fds) {
    if (fd >= 0) ::close(fd);
  }
  fds.clear();
}
}  // namespace

ReceivedSurfacePacket::ReceivedSurfacePacket(nlohmann::json message,
                                             std::vector<int> fds)
    : message(std::move(message)), fds(std::move(fds)) {}

ReceivedSurfacePacket::~ReceivedSurfacePacket() { close_all(fds); }

ReceivedSurfacePacket::ReceivedSurfacePacket(ReceivedSurfacePacket&& other) noexcept
    : message(std::move(other.message)), fds(std::move(other.fds)) {
  other.fds.clear();
}

ReceivedSurfacePacket& ReceivedSurfacePacket::operator=(
    ReceivedSurfacePacket&& other) noexcept {
  if (this == &other) return *this;
  close_all(fds);
  message = std::move(other.message);
  fds = std::move(other.fds);
  other.fds.clear();
  return *this;
}

SurfaceChannel::SurfaceChannel(int fd, bool owns_fd)
    : fd_(fd), owns_fd_(owns_fd) {
  if (fd < 0) throw std::invalid_argument("surface channel fd is invalid");
}

SurfaceChannel::~SurfaceChannel() {
  if (owns_fd_ && fd_ >= 0) ::close(fd_);
}

SurfaceChannel::SurfaceChannel(SurfaceChannel&& other) noexcept
    : fd_(std::exchange(other.fd_, -1)), owns_fd_(other.owns_fd_) {}

SurfaceChannel& SurfaceChannel::operator=(SurfaceChannel&& other) noexcept {
  if (this == &other) return *this;
  if (owns_fd_ && fd_ >= 0) ::close(fd_);
  fd_ = std::exchange(other.fd_, -1);
  owns_fd_ = other.owns_fd_;
  return *this;
}

std::optional<SurfaceChannel> SurfaceChannel::from_environment() {
  const char* value = std::getenv("LOCALBOORU_SURFACE_FD");
  if (value == nullptr || *value == '\0') return std::nullopt;
  char* end = nullptr;
  errno = 0;
  const long parsed = std::strtol(value, &end, 10);
  if (errno != 0 || end == value || *end != '\0' || parsed < 0 ||
      parsed > std::numeric_limits<int>::max()) {
    throw std::runtime_error("LOCALBOORU_SURFACE_FD is invalid");
  }
  return SurfaceChannel(static_cast<int>(parsed));
}

void SurfaceChannel::send(const nlohmann::json& message,
                          const std::vector<int>& fds) const {
  if (fds.size() > kMaxMessageFds) {
    throw std::invalid_argument("surface message has too many file descriptors");
  }
  const std::string payload = message.dump();
  if (payload.size() > kMaxMessageBytes) {
    throw std::invalid_argument("surface message exceeds bounded payload size");
  }
  iovec iov{const_cast<char*>(payload.data()), payload.size()};
  std::array<std::byte, CMSG_SPACE(kMaxMessageFds * sizeof(int))> control{};
  msghdr header{};
  header.msg_iov = &iov;
  header.msg_iovlen = 1;
  if (!fds.empty()) {
    header.msg_control = control.data();
    header.msg_controllen = CMSG_SPACE(fds.size() * sizeof(int));
    cmsghdr* cmsg = CMSG_FIRSTHDR(&header);
    cmsg->cmsg_level = SOL_SOCKET;
    cmsg->cmsg_type = SCM_RIGHTS;
    cmsg->cmsg_len = CMSG_LEN(fds.size() * sizeof(int));
    std::memcpy(CMSG_DATA(cmsg), fds.data(), fds.size() * sizeof(int));
  }
  const ssize_t sent = ::sendmsg(fd_, &header, MSG_NOSIGNAL);
  if (sent < 0) throw system_error("failed to send surface packet");
  if (static_cast<std::size_t>(sent) != payload.size()) {
    throw std::runtime_error("surface channel sent a partial packet");
  }
}

ReceivedSurfacePacket SurfaceChannel::receive() const {
  auto packet = receive_with_flags(0);
  if (!packet) {
    throw std::logic_error("blocking surface receive returned no packet");
  }
  return std::move(*packet);
}

std::optional<ReceivedSurfacePacket> SurfaceChannel::try_receive() const {
  return receive_with_flags(MSG_DONTWAIT);
}

std::optional<ReceivedSurfacePacket> SurfaceChannel::receive_with_flags(
    int flags) const {
  std::array<char, kMaxMessageBytes> payload{};
  std::array<std::byte, CMSG_SPACE(kMaxMessageFds * sizeof(int))> control{};
  iovec iov{payload.data(), payload.size()};
  msghdr header{};
  header.msg_iov = &iov;
  header.msg_iovlen = 1;
  header.msg_control = control.data();
  header.msg_controllen = control.size();
  const ssize_t received = ::recvmsg(fd_, &header, MSG_CMSG_CLOEXEC | flags);
  if (received < 0 && flags != 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) {
    return std::nullopt;
  }
  if (received < 0) throw system_error("failed to receive surface packet");
  if (received == 0) throw std::runtime_error("surface channel closed");
  if ((header.msg_flags & (MSG_TRUNC | MSG_CTRUNC)) != 0) {
    throw std::runtime_error("surface channel packet or descriptors truncated");
  }

  std::vector<int> fds;
  for (cmsghdr* cmsg = CMSG_FIRSTHDR(&header); cmsg != nullptr;
       cmsg = CMSG_NXTHDR(&header, cmsg)) {
    if (cmsg->cmsg_level != SOL_SOCKET || cmsg->cmsg_type != SCM_RIGHTS) continue;
    const std::size_t bytes = cmsg->cmsg_len - CMSG_LEN(0);
    const std::size_t count = bytes / sizeof(int);
    const int* received_fds = reinterpret_cast<const int*>(CMSG_DATA(cmsg));
    fds.insert(fds.end(), received_fds, received_fds + count);
  }
  if (fds.size() > kMaxMessageFds) {
    close_all(fds);
    throw std::runtime_error("surface channel received too many descriptors");
  }
  try {
    return ReceivedSurfacePacket(
        nlohmann::json::parse(payload.data(), payload.data() + received),
        std::move(fds));
  } catch (...) {
    close_all(fds);
    throw;
  }
}

}  // namespace localbooru::native_video
